File: sanity_test.py
#!/usr/bin/env python3
"""
Sanity Test for GitHub Copilot Functionality

This file demonstrates and tests various GitHub Copilot capabilities including:
- Code completion and suggestions
- Function generation from comments
- Bug detection and fixes
- Test case generation
- Documentation generation
"""

import os
import sys
import json
from datetime import datetime
from typing import List, Dict, Optional


class SanityTest:
    """Main class for running Copilot sanity tests."""
    
    def __init__(self):
        self.test_results = []
        self.start_time = datetime.now()
    
    def run_all_tests(self) -> bool:
        """Run all sanity tests and return True if all pass."""
        print("🚀 Starting GitHub Copilot Sanity Tests...")
        
        tests = [
            self.test_basic_functionality,
            self.test_data_processing,
            self.test_algorithm_implementation,
            self.test_error_handling,
            self.test_documentation_generation
        ]
        
        for test in tests:
            try:
                result = test()
                self.test_results.append({
                    'test': test.__name__,
                    'status': 'PASS' if result else 'FAIL',
                    'timestamp': datetime.now().isoformat()
                })
                status_emoji = "✅" if result else "❌"
                print(f"{status_emoji} {test.__name__}: {'PASSED' if result else 'FAILED'}")
            except Exception as e:
                self.test_results.append({
                    'test': test.__name__,
                    'status': 'ERROR',
                    'error': str(e),
                    'timestamp': datetime.now().isoformat()
                })
                print(f"💥 {test.__name__}: ERROR - {e}")
        
        return all(result['status'] == 'PASS' for result in self.test_results)
    
    def test_basic_functionality(self) -> bool:
        """Test basic Python functionality and Copilot suggestions."""
        # Test basic arithmetic - Copilot should suggest mathematical operations
        numbers = [1, 2, 3, 4, 5]
        total = sum(numbers)
        average = total / len(numbers)
        
        # Test list comprehension - Copilot should suggest Pythonic patterns
        squared_numbers = [n ** 2 for n in numbers]
        even_numbers = [n for n in numbers if n % 2 == 0]
        
        return (total == 15 and 
                average == 3.0 and 
                squared_numbers == [1, 4, 9, 16, 25] and
                even_numbers == [2, 4])
    
    def test_data_processing(self) -> bool:
        """Test data processing capabilities that Copilot might suggest."""
        # Sample data for processing
        sample_data = [
            {"name": "Alice", "age": 30, "city": "New York"},
            {"name": "Bob", "age": 25, "city": "San Francisco"},
            {"name": "Charlie", "age": 35, "city": "Chicago"},
            {"name": "Diana", "age": 28, "city": "New York"}
        ]
        
        # Filter data - Copilot should suggest filtering patterns
        ny_residents = [person for person in sample_data if person["city"] == "New York"]
        adults_over_30 = [person for person in sample_data if person["age"] > 30]
        
        # Sort data - Copilot should suggest sorting operations
        sorted_by_age = sorted(sample_data, key=lambda x: x["age"])
        
        return (len(ny_residents) == 2 and 
                len(adults_over_30) == 1 and
                sorted_by_age[0]["name"] == "Bob")
    
    def test_algorithm_implementation(self) -> bool:
        """Test algorithm implementation that Copilot might help with."""
        # Fibonacci sequence - common algorithm Copilot suggests
        def fibonacci(n: int) -> int:
            if n <= 1:
                return n
            return fibonacci(n - 1) + fibonacci(n - 2)
        
        # Binary search - another common algorithm
        def binary_search(arr: List[int], target: int) -> int:
            left, right = 0, len(arr) - 1
            while left <= right:
                mid = (left + right) // 2
                if arr[mid] == target:
                    return mid
                elif arr[mid] < target:
                    left = mid + 1
                else:
                    right = mid - 1
            return -1
        
        # Test the algorithms
        fib_result = fibonacci(6)  # Should be 8
        search_result = binary_search([1, 3, 5, 7, 9, 11], 7)  # Should be index 3
        
        return fib_result == 8 and search_result == 3
    
    def test_error_handling(self) -> bool:
        """Test error handling patterns that Copilot might suggest."""
        def safe_divide(a: float, b: float) -> Optional[float]:
            """Safely divide two numbers with error handling."""
            try:
                if b == 0:
                    raise ValueError("Division by zero is not allowed")
                return a / b
            except (TypeError, ValueError) as e:
                print(f"Error in division: {e}")
                return None
        
        def safe_file_read(filename: str) -> Optional[str]:
            """Safely read a file with error handling."""
            try:
                with open(filename, 'r') as file:
                    return file.read()
            except FileNotFoundError:
                return None
            except PermissionError:
                return None
        
        # Test error handling
        valid_division = safe_divide(10, 2)  # Should be 5.0
        invalid_division = safe_divide(10, 0)  # Should be None
        missing_file = safe_file_read("nonexistent_file.txt")  # Should be None
        
        return (valid_division == 5.0 and 
                invalid_division is None and 
                missing_file is None)
    
    def test_documentation_generation(self) -> bool:
        """Test documentation patterns that Copilot might generate."""
        def calculate_statistics(numbers: List[float]) -> Dict[str, float]:
            """
            Calculate basic statistics for a list of numbers.
            
            Args:
                numbers: A list of numeric values
                
            Returns:
                A dictionary containing mean, median, and standard deviation
                
            Raises:
                ValueError: If the input list is empty
            """
            if not numbers:
                raise ValueError("Cannot calculate statistics for empty list")
            
            # Calculate mean
            mean = sum(numbers) / len(numbers)
            
            # Calculate median
            sorted_numbers = sorted(numbers)
            n = len(sorted_numbers)
            median = (sorted_numbers[n//2] if n % 2 == 1 
                     else (sorted_numbers[n//2-1] + sorted_numbers[n//2]) / 2)
            
            # Calculate standard deviation
            variance = sum((x - mean) ** 2 for x in numbers) / len(numbers)
            std_dev = variance ** 0.5
            
            return {
                "mean": mean,
                "median": median,
                "std_dev": std_dev
            }
        
        # Test the function
        test_numbers = [1, 2, 3, 4, 5]
        stats = calculate_statistics(test_numbers)
        
        return (abs(stats["mean"] - 3.0) < 0.001 and
                abs(stats["median"] - 3.0) < 0.001 and
                abs(stats["std_dev"] - 1.414) < 0.01)
    
    def generate_report(self) -> str:
        """Generate a test report."""
        end_time = datetime.now()
        duration = (end_time - self.start_time).total_seconds()
        
        passed = sum(1 for result in self.test_results if result['status'] == 'PASS')
        total = len(self.test_results)
        
        report = f"""
=== GitHub Copilot Sanity Test Report ===
Start Time: {self.start_time.isoformat()}
End Time: {end_time.isoformat()}
Duration: {duration:.2f} seconds
Tests Passed: {passed}/{total}
Success Rate: {(passed/total)*100:.1f}%

Detailed Results:
"""
        
        for result in self.test_results:
            report += f"- {result['test']}: {result['status']}"
            if 'error' in result:
                report += f" ({result['error']})"
            report += "\n"
        
        return report


def main():
    """Main function to run the sanity tests."""
    print("GitHub Copilot Sanity Test Suite")
    print("=" * 40)
    
    # Create and run tests
    sanity_test = SanityTest()
    success = sanity_test.run_all_tests()
    
    # Generate and display report
    report = sanity_test.generate_report()
    print(report)
    
    # Save report to file
    with open("sanity_test_report.txt", "w") as f:
        f.write(report)
    
    # Exit with appropriate code
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: sanity_test.js
#!/usr/bin/env node
/**
 * JavaScript Sanity Test for GitHub Copilot
 * 
 * This file demonstrates Copilot capabilities in JavaScript including:
 * - Modern ES6+ syntax suggestions
 * - Async/await patterns
 * - Error handling
 * - DOM manipulation concepts
 * - API interaction patterns
 */

class JavaScriptSanityTest {
    constructor() {
        this.testResults = [];
        this.startTime = new Date();
    }

    /**
     * Run all JavaScript sanity tests
     * @returns {Promise<boolean>} True if all tests pass
     */
    async runAllTests() {
        console.log('🚀 Starting JavaScript Copilot Sanity Tests...');
        
        const tests = [
            this.testES6Features,
            this.testAsyncOperations,
            this.testArrayMethods,
            this.testObjectManipulation,
            this.testErrorHandling
        ];

        for (const test of tests) {
            try {
                const result = await test.call(this);
                this.testResults.push({
                    test: test.name,
                    status: result ? 'PASS' : 'FAIL',
                    timestamp: new Date().toISOString()
                });
                const statusEmoji = result ? "✅" : "❌";
                console.log(`${statusEmoji} ${test.name}: ${result ? 'PASSED' : 'FAILED'}`);
            } catch (error) {
                this.testResults.push({
                    test: test.name,
                    status: 'ERROR',
                    error: error.message,
                    timestamp: new Date().toISOString()
                });
                console.log(`💥 ${test.name}: ERROR - ${error.message}`);
            }
        }

        return this.testResults.every(result => result.status === 'PASS');
    }

    /**
     * Test ES6+ features that Copilot commonly suggests
     */
    testES6Features() {
        // Arrow functions
        const add = (a, b) => a + b;
        const multiply = (a, b) => a * b;

        // Template literals
        const name = "Copilot";
        const greeting = `Hello, ${name}! Welcome to the sanity test.`;

        // Destructuring
        const person = { firstName: "John", lastName: "Doe", age: 30 };
        const { firstName, lastName, age } = person;
        const fullName = `${firstName} ${lastName}`;

        // Spread operator
        const numbers = [1, 2, 3];
        const moreNumbers = [...numbers, 4, 5, 6];
        const maxNumber = Math.max(...moreNumbers);

        // Default parameters
        const greetUser = (name = "User", greeting = "Hello") => `${greeting}, ${name}!`;

        return (
            add(2, 3) === 5 &&
            multiply(4, 5) === 20 &&
            greeting.includes("Copilot") &&
            fullName === "John Doe" &&
            age === 30 &&
            moreNumbers.length === 6 &&
            maxNumber === 6 &&
            greetUser() === "Hello, User!" &&
            greetUser("Alice") === "Hello, Alice!"
        );
    }

    /**
     * Test asynchronous operations and promises
     */
    async testAsyncOperations() {
        // Promise-based delay function
        const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

        // Async function with error handling
        const fetchData = async (url) => {
            try {
                // Simulate API call
                await delay(10);
                return { data: "mock data", status: 200 };
            } catch (error) {
                throw new Error(`Failed to fetch data: ${error.message}`);
            }
        };

        // Promise.all for concurrent operations
        const concurrentOperations = async () => {
            const promises = [
                fetchData("api/users"),
                fetchData("api/posts"),
                fetchData("api/comments")
            ];
            return await Promise.all(promises);
        };

        try {
            const singleResult = await fetchData("api/test");
            const multipleResults = await concurrentOperations();

            return (
                singleResult.status === 200 &&
                singleResult.data === "mock data" &&
                multipleResults.length === 3 &&
                multipleResults.every(result => result.status === 200)
            );
        } catch (error) {
            return false;
        }
    }

    /**
     * Test array methods that Copilot frequently suggests
     */
    testArrayMethods() {
        const numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
        const fruits = ["apple", "banana", "orange", "grape"];

        // Filter methods
        const evenNumbers = numbers.filter(n => n % 2 === 0);
        const longFruits = fruits.filter(fruit => fruit.length > 5);

        // Map methods
        const squaredNumbers = numbers.map(n => n ** 2);
        const uppercaseFruits = fruits.map(fruit => fruit.toUpperCase());

        // Reduce methods
        const sum = numbers.reduce((acc, n) => acc + n, 0);
        const product = numbers.slice(0, 5).reduce((acc, n) => acc * n, 1);

        // Find methods
        const firstEven = numbers.find(n => n % 2 === 0);
        const indexOfOrange = fruits.findIndex(fruit => fruit === "orange");

        // Some and every
        const hasEvenNumbers = numbers.some(n => n % 2 === 0);
        const allPositive = numbers.every(n => n > 0);

        return (
            evenNumbers.length === 5 &&
            longFruits.length === 2 &&  // Fixed: both "banana" and "orange" have length 6
            longFruits.includes("banana") &&
            squaredNumbers[2] === 9 &&
            uppercaseFruits[0] === "APPLE" &&
            sum === 55 &&
            product === 120 &&
            firstEven === 2 &&
            indexOfOrange === 2 &&
            hasEvenNumbers === true &&
            allPositive === true
        );
    }

    /**
     * Test object manipulation patterns
     */
    testObjectManipulation() {
        // Object creation and manipulation
        const createUser = (name, email, age) => ({
            name,
            email,
            age,
            isActive: true,
            lastLogin: new Date(),
            getDisplayName() {
                return `${this.name} (${this.email})`;
            }
        });

        // Object methods and property access
        const user = createUser("Alice Smith", "alice@example.com", 28);
        const displayName = user.getDisplayName();

        // Object destructuring and cloning
        const { name, email, age } = user;
        const userCopy = { ...user, age: 29 };

        // Object keys and values
        const userKeys = Object.keys(user);
        const hasNameProperty = userKeys.includes('name');

        // Dynamic property access
        const propertyName = 'email';
        const emailValue = user[propertyName];

        return (
            user.name === "Alice Smith" &&
            user.isActive === true &&
            displayName.includes("Alice Smith") &&
            userCopy.age === 29 &&
            user.age === 28 &&
            hasNameProperty === true &&
            emailValue === "alice@example.com"
        );
    }

    /**
     * Test error handling patterns
     */
    testErrorHandling() {
        // Try-catch with custom errors
        const validateAge = (age) => {
            if (typeof age !== 'number') {
                throw new TypeError('Age must be a number');
            }
            if (age < 0 || age > 150) {
                throw new RangeError('Age must be between 0 and 150');
            }
            return age;
        };

        // Error handling with promises
        const asyncValidation = async (data) => {
            try {
                if (!data) {
                    throw new Error('Data is required');
                }
                return { valid: true, data };
            } catch (error) {
                return { valid: false, error: error.message };
            }
        };

        // Test error scenarios
        let results = {
            validAge: false,
            invalidTypeAge: false,
            invalidRangeAge: false,
            validAsyncData: false,
            invalidAsyncData: false
        };

        try {
            validateAge(25);
            results.validAge = true;
        } catch (error) {
            // Should not reach here
        }

        try {
            validateAge("invalid");
        } catch (error) {
            results.invalidTypeAge = error instanceof TypeError;
        }

        try {
            validateAge(-5);
        } catch (error) {
            results.invalidRangeAge = error instanceof RangeError;
        }

        // Test async validation
        return asyncValidation({ test: "data" })
            .then(result => {
                results.validAsyncData = result.valid;
                return asyncValidation(null);
            })
            .then(result => {
                results.invalidAsyncData = !result.valid;
                return Object.values(results).every(Boolean);
            });
    }

    /**
     * Generate a comprehensive test report
     */
    generateReport() {
        const endTime = new Date();
        const duration = (endTime - this.startTime) / 1000;
        
        const passed = this.testResults.filter(result => result.status === 'PASS').length;
        const total = this.testResults.length;
        
        let report = `
=== JavaScript Copilot Sanity Test Report ===
Start Time: ${this.startTime.toISOString()}
End Time: ${endTime.toISOString()}
Duration: ${duration.toFixed(2)} seconds
Tests Passed: ${passed}/${total}
Success Rate: ${((passed/total)*100).toFixed(1)}%

Detailed Results:
`;
        
        this.testResults.forEach(result => {
            report += `- ${result.test}: ${result.status}`;
            if (result.error) {
                report += ` (${result.error})`;
            }
            report += '\n';
        });
        
        return report;
    }
}

/**
 * Main function to run the JavaScript sanity tests
 */
async function main() {
    console.log('JavaScript GitHub Copilot Sanity Test Suite');
    console.log('='.repeat(45));
    
    const sanityTest = new JavaScriptSanityTest();
    const success = await sanityTest.runAllTests();
    
    const report = sanityTest.generateReport();
    console.log(report);
    
    // Write report to file if running in Node.js environment
    if (typeof require !== 'undefined') {
        const fs = require('fs');
        fs.writeFileSync('js_sanity_test_report.txt', report);
    }
    
    process.exit(success ? 0 : 1);
}

// Run if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { JavaScriptSanityTest };